### PR TITLE
Fix `persistQuery` option for push, replace actions

### LINKED ISSUE #224: `persistQuery` option doesn't work with `push` action creator
As alluded to in #172, it looks like the `persistQuery` option doesn't work when you dispatch a `push` action directly.

From what I can tell, when you dispatch `push` directly, it immediately hits the middleware, which updates the url immediately with the unmerged query. _After_ this happens, the action hits the reducer, which then merges the existing query params with the new ones and updates the redux store.

I'm not familiar enough yet with Redux to know what the correct thing to do here is except make the middleware do the merging before calling any `history` methods.

### PR DESCRIPTION
**Pull request by adamesque**
This PR adds awareness of `persistQuery` to the middleware, so that when we push/replace a new history state, the browser's URL includes the merged set of query params.

It also adds a new utility function, `mergeQueries`, which takes two query objects, merges them, and return a `Location`-compatible object with the merged queries and a merged `search` string.

Let me know what you think!

Fixes #224

### CODE REVIEW

**Comment by PeterLeeHyfn** [comment]
any updates on this PR?

**Comment by adamesque** [comment]
@PeterLeeHyfn I've been happily using my fork with this fix. This project appears to be dormant right now; seems like it's maintained by a single person & I don't begrudge them taking a break; open source takes an incredible amount of energy. If I had the bandwidth I'd offer to help maintain, but I can't really make that offer in good faith right now.

**Comment by ryan-roemer** [comment]
Hi @adamesque ! 

This one is a priority of ours and we're aiming to get back to the backlog soon. Biggest blocker is a small child cramping our main maintainer's style + we haven't been able to train up some other folks to better help maintain the repo.

It's on the radar, just not right now. Totally understand needing to get by with a fork in the meantime, and appreciate any patience y'all have for us as we try and get more support ❤️  going for this repo....

File: src/util/merge-queries.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// @flow
+import type { Location, Query } from '../types';
+
+import qs from 'query-string';
+
+export default (oldQuery: Query = {}, newQuery: Query = {}): Location => {
```
Comment (aweary) [inline src/util/merge-queries.js]: We can avoid the default params in this case, object spread works fine with `undefined`

File: test/util/merge-queries.spec.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { expect } from 'chai';
+import mergeQueries from '../../src/util/merge-queries';
+
+describe('mergeQueries', () => {
+  it('merges two query objects and adds a merged search string', () => {
```
Comment (aweary) [inline test/util/merge-queries.spec.js]: Let's add a test for when only a single query exists.

**Comment by adamesque** [comment]
Thanks for the review, @aweary! I pushed up a few commits to address your feedback.

**Comment by yogaboll** [comment]
It would be great if we could get this pull request merged and published soon. It's really needed.

**Comment by ryan-roemer** [comment]
@yogaboll -- I think we have just one more test request from @adamesque to go. If you want to pull this branch, create a test and just paste it here in a comment, we might able to go forward if @adamesque doesn't have time right now.

**Comment by aweary** [comment]
@ryan-roemer looks like that test was added (Github just didn't collapse my comment). @tptee if this looks good to you then it should be good to merge 👍

**Comment by tptee** [comment]
Out in `v14.2.2`!

**Pull request merged**